v3.1.3 patches

cary-ilm opened the pull request:
Cherry-picked commits for v3.1.3. See CHANGES.md for the list of PR's.

This is a first pass, we can add further changes to it before merging.

Also, we agreed to bend the libtool versioning rules and call the library version 30.3.0 (v3.1.2 was 30.2.0), even though strictly speaking the revision should be bumped because we added to the API. But v3.1.3 should be drop-in compatible with v3.1.2 (even with the compression level changes).

kdt3rd committed:
Fix ECX register access

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix when compiling against pre-3.1 version of Imath (#1145)

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Enable deprecating attributes, deprecate dwaCompression

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add ephemeral settings for compression levels, propagate to compressors

NB: This changes the ABI of ImfHeader, so must be a 3.2 feature

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add ephemeral compression level settings to core

with unit tests

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Enable Google OSS Fuzz to also test Core library (#1150)

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add layer to avoid changing ABI such that this can be ported to 3.1

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add global setting of the default compression levels

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix issue with static libraries and order of deletion on MacOS

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
add matching functions to control global default zip / dwa levels to C core

Also adds the mechanism to the context initializer

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Alternate implementation to attempt to fix MacOS and Windows at the same time

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add extra safety checks, proper ifdef checks against threading enabled

Also, comment the more complex workaround to get to a stable static
destruction state when compiling to static libraries

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

peterhillman committed:
validate filesize before allocating chunk memory (#1161)

Signed-off-by: Peter Hillman <peterh@wetafx.co.nz>

peterhillman committed:
validate dwaCompressionLevel attribute type (#1160)

Signed-off-by: Peter Hillman <peterh@wetafx.co.nz>

peterhillman committed:
don't validate chunk size when file_size unknown (#1163)

Signed-off-by: Peter Hillman <peterh@wetafx.co.nz>

kdt3rd committed:
Also fail when the user provides a preview image that has a zero size coordinate (#1164)

oss fuzz 39399

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Prevent read when offset past the end of the memstream (#1165)

normal file semantics will return 0 for this indicating EOF

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

peterhillman committed:
fix int overflow in calc_level_size (#1166)

Signed-off-by: Peter Hillman <peterh@wetafx.co.nz>

peterhillman committed:
Fix loop iterators in ImfCheckFile.cpp (#1167)

* fix dataWindow width signed calc
* use int64_t in CheckFile loop to prevent wraparound
* tile indexes should start from 0 not datawin.min

Signed-off-by: Peter Hillman <peterh@wetafx.co.nz>

kdt3rd committed:
Clean up error messages, check against packed size of 0, integer overflow (#1169)

OSS Fuzz 39522, 39526, 39529

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Refactor attribute size checks (#1168)

* Avoid uninitialized memory error in non-sanitized libstdc++

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

* Refactor attribute size checks

for large attribute chunks, clear memory earlier, centralize file size
check, fix uninitialized memory error with memory sanitizer

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

* Fix typo

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

peterhillman committed:
fix casts in readUInt shifts (#1172)

Signed-off-by: Peter Hillman <peterh@wetafx.co.nz>

aras-p committed:
Zip: switch to compression level 4 instead of default 6 (#1125)

In my test of 18 various exr files (total raw data size 1057MB):
- zip 6: 2.452x ratio, 206MB/s compression
- zip 4: 2.421x ratio, 437MB/s compression

So a tiny drop of compression ratio, but compression is more than 2x faster. This makes writing zip faster than writing uncompressed (386MB/s). Decompression speed unaffected.

Signed-off-by: Aras Pranckevicius <aras@unity3d.com>

kdt3rd committed:
Fix misc issues due to OSS-fuzz (#1173)

* Fix issues with fuzz test driver

- Adds a (disabled) error handler print to make th test cases easier to
part
- Fixes an initialization of buffers to use the better half
- mark a couple of functions internal (static)

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

* Add chunk for offsets in chunk table to be (somewhat validate)

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

* Clarify comment, simplify logic check

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

* Add check in mirror decode pipeline for alloc of 0 bytes

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

* Adjust logic so we can handle single scanline compression with image sampling

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

* switch to controlling prints for checking behind a getenv

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

cary-ilm committed:
Update docs with link to EasyCLA (#1174)

* Remove local copies of outdated CLA's

Signed-off-by: Cary Phillips <cary@ilm.com>

peterhillman committed:
use std::abs in chromaticity sanity tests (fixes #1177) (#1178)

Signed-off-by: Peter Hillman <peterh@wetafx.co.nz>

cary-ilm committed:
Release notes for v3.1.3

Signed-off-by: Cary Phillips <cary@ilm.com>

cary-ilm committed:
Update version to 3.1.3 and lib version to 30.3.0

Signed-off-by: Cary Phillips <cary@ilm.com>

xlietz committed:
Update CI builds: (#1176)

- Multiple Linux builds only for VFXP-2021 (current).
- Single Linux build for other VFX platform years.
- Add Linux legacy C++11 build.
- Update Mac and Windows builds for VFXP-2021/2022

Signed-off-by: Christina Tempelaar-Lietz <xlietz@gmail.com>

lgritz commented:
Hi, please don't release until we get a fix for the broken deep reads with the core library that was caused by #1169.

cary-ilm commented:
Yes, just getting it started.

kdt3rd committed:
Require at least one channel (#1182)

When we started core, there was a thought to allow an empty image to
enable a metadata part. However, this needs wider support and causes
regressions from the existing c++ behavior

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix test for bad chunk data to allow for 0-sample deep chunks (#1185)

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix decoding of piz when y sampling is not the same for all channels (#1183)

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fixes an issue computing the unpacked size of a chunk (#1184)

This fixes an issue when the y_sampling is larger than the height of a
chunk (and the height is not 1). Further, adds casts, use larger type to
avoid overflow

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add extra tile validation (#1186)


Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add size check to memory stream check program (#1187)

Accidentally giving a folder name to the program exposed that the c++
library returns very large results for seeking to the end of a directory
file node, handle and return quickly

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

peterhillman committed:
reduce iterations in testIDManifest to speed up (#1180)

Signed-off-by: Peter Hillman <peterh@wetafx.co.nz>

cary-ilm committed:
Update release notes with recent PR's and OSS-fuzz issues

Signed-off-by: Cary Phillips <cary@ilm.com>

meshula reviewed (approved):
All of these cherry-picks seem appropriate to the release.
